src/main/java/com/puppycrawl/tools/checkstyle/Checker.java

```diff
@@ -138,7 +138,7 @@ public class Checker extends AutomaticBean implements MessageDispatcher, RootMod
      * The instance needs to be contextualized and configured.
      */
     public Checker() {
-        addListener(counter);
+        listeners.add(counter);
```

Comment:
We won't have the childContext when this is called since we haven't read in the config yet, so I feel it is better to just add the listener directly like this as it will not need any of that information. This will clearly show it is something different.

Reply:
We should similarly be doing the same with https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/Checker.java#L486-L498 as we need to differentiate items being brought in from the config versus those being added from outside forces through our `add*` methods in `Checker`. I am starting to think those add methods should be removed as they seem more for test only.

Reply:
I do not understand how this will make any difference in this case. Do we have a test that will fail without directly accessing collections?

Reply:
https://github.com/rnveach/checkstyle/commit/9a5598a18525509a6467e3138d5b5d6d7df246f9#diff-650e834bca871d6bf6c1b3800ff4116b2d7ab4fe5de6489a042652c2f72e024dR555-R558
https://github.com/rnveach/checkstyle/commit/9a5598a18525509a6467e3138d5b5d6d7df246f9#diff-650e834bca871d6bf6c1b3800ff4116b2d7ab4fe5de6489a042652c2f72e024dL501-R504

The `add` methods will be changed to methods for outside us only. This is to prevent accidentally `contextualize`ing the object twice, once inside the `setupChild` method and again inside `add`. If it is something this class absolutely needs, then it should handle all its needs itself instead of mixing it up with outside forces.

The way I currently coded it, there will be no failures, but improper use will lead to things not working like if they expect the locale, hence why this PR is titled "order of execution".

Reply:
Imagine a bank can have money deposited through a teller or through an armored car which the bank owns. You will want different checks for each. The personnel depositing money through the teller is fraud in some way. On the other end, you will want to make sure the armored car has the right credentials to be let into the vault to directly deposit.

I am adding this separation here to ensure there are no mistakes by confusing layers. Luckily, most of this being done is just very simple methods that have no big impact.

Reply:
What will prevent some future bank manager from sending the armored car through the teller? Especially since the interface specifies this?

Reply:
There is nothing. How can i tell a class to not use it's own public method? I am open to suggestions.

Reply:
> I am starting to think those add methods should be removed as they seem more for test only, but it seems maven-checkstyle uses one of them.

we support Bean based configuration of Checkstyle to let users use us without XML, for example Sonar plugin or any other plugin with different persistence of module configs.